Add option for "clean edges" to belt_balancer

ignaloidas opened the pull request:
Allows for checking for smallest balancers with arbitrary input configuration - without this option, some weirdness with splitter colors disallows splitters directly outputing onto the grid edge.

Was checking for some minimal sized balancers with no restriction on input/output configuration, and came across this problem.

ignaloidas committed:
Add option for "clean edges" to belt_balancer

Allows for checking for smallest balancers with arbitrary input configuration - without this option, some weirdness with splitter colors disallows splitters directly outputing onto the grid edge.

R-O-C-K-E-T commented:
Since splitter tiles inherit colour from their input, it make sense that splitters wouldn't work for outputs. Taking that into consideration, it seems like enabling this new `--clean-edges` flag simply fixes a bug in the `--custom` mode. Where not all valid solutions were being recognised. Hence I think this should be baseline functionality instead of a flag.

Making this change would fix up a consistency issue, where `--custom` used any non-empty tile for inputs/outputs. While the other modes (default, --90, --180) only used belts.
For a bit more flexibility, you might want to investigate adding support for `--use-ends` with `--custom`.

R-O-C-K-E-T commented on factorio_sat/belt_balancer.py:
```diff
@@ -291,6 +291,23 @@ def set_nonempty_tiles(grid: Grid, blueprint_or_json: str):
+    for x in range(grid.width):
+        tile_a = grid.get_tile_instance(x, 0)
+        tile_b = grid.get_tile_instance(x, grid.height - 1)
+        grid.clauses += [[-tile_a.is_splitter], [-tile_a.is_splitter_head], [-tile_a.is_underground_in], [-tile_a.is_underground_out]]
```
Might be clearer and more future-proof to write this as edge tiles are either empty or belt. Instead of enumerating all non-belt types. 
```python
grid.clauses.append([tile_a.is_empty, tile_a.is_belt])
grid.clauses.append([tile_b.is_empty, tile_b.is_belt])
```

ignaloidas committed:
Restrict input/output tiles to belts only

Adjust tests accordingly, update expand-underground optimization to not
expand into input/output belts

ignaloidas commented:
> Since splitter tiles inherit colour from their input, it make sense that splitters wouldn't work for outputs. Taking that into consideration, it seems like enabling this new `--clean-edges` flag simply fixes a bug in the `--custom` mode. Where not all valid solutions were being recognised. Hence I think this should be baseline functionality instead of a flag.
>
> Making this change would fix up a consistency issue, where `--custom` used any non-empty tile for inputs/outputs. While the other modes (default, --90, --180) only used belts.

Oh, I missed a fix I did in my local branch for testing out some weird changes, where inputs/outputs are disallowed from being anything but a belt, copied that change into this. Though even then, the problem still exists, because input/output usage would be inset one from the edge with the edge tiles used, which doesn't quite match up with how most people count balancer size.

>For a bit more flexibility, you might want to investigate adding support for `--use-ends` with `--custom`.

I'm not entirely certain if this should be merged with `--use-ends`? The default behavior without `--custom` is to allow usage of the edges, and if the edges are disallowed by default, then there is some potential for confusion, because you now need to pass in dimensions that have 2 tiles of padding for the edges, making some trivial balancers unable to fit, whereas if you ran it without `--custom`, it would work. Which in my mind doesn't make sense, because I see `--custom` as removing most restriction, and adding it shouldn't make a command that outputs a balancer no longer do so.

ignaloidas commented:
> Oh, I missed a fix I did in my local branch for testing out some weird changes, where inputs/outputs are disallowed from being anything but a belt, copied that change into this.

Oh, seems like a bunch of tests assume that inputs/outputs can be non-belt, will go over them to fix them up

ignaloidas commented:
Fixed the tests, and set the `--expand-undergrounds` optimization to ignore the possibility of expanding into input/output belts
